### DISCUSSION TITLE
Nested `Mount` and `Route`

### DISCUSSION THREAD

**Discussion by edthrn** [encode/starlette#2802] [Q&A] [answered]
I'd like to use a pattern where nested `Mount()` allow me to have a nicely factorize routing table, as such:

```python
from starlette.applications import Starlette    
from starlette.routing import Route, Mount

routes = [
    Mount(
        "/api",
        routes=[
            Route("/foos", list_foos),
	    Mount(
		"/foo/{foo_id:uuid}",
		routes=[
		    Route("/bars", list_bars),
		    Route("/bar/{bar_id:uuid}", get_bar),
		    Route("/quuxs", list_quuxs),
		    Route("/quux/{quux_id:uuid}", get_quux),
		    Route("/quux/{quux_id:uuid}/bazs", list_bazs),
		    Mount(
		        "/monitoring",
		        routes=[
		            Mount(
		                "/bar/{bar_id:uuid}",
		                routes=[
		                    Route("/overview", get_overview),
		                    Route("/quuxs", list_quuxs_monitoring),
		                    Route("/bazs", list_bazs_monitoring),
		                ],
		            ),
		        ],
		    ),
		],
		middleware=[Middleware(FooMiddleware)],
	    ),
            Route("/support", get_support),
        ],
        middleware=[
            Middleware(APIMiddleware1),
       ],
    )
]

app = Starlette(routes=routes, debug=True)
```

Serving this app, `GET /api/foos` works well, but `GET /api/foo/{foo_id}/bars` returns a 404, suggesting that nested mounts don't work as I expected.

I tried something similar using the lower-level `Router` object, but still no luck.

Am I missing something?

---
_Note: I stumbled upon [this thread](https://github.com/encode/starlette/issues/380) and [this one](https://github.com/encode/starlette/issues/869) but they don't seem related to my issue._

**Comment by edthrn** [comment]
I managed to make it work... It seems like the problem comes from the `Mount("/foo/{foo_id:uuid}")` part.

- ✅ This works: `GET api/foo/74180c32-cc6f-47b6-9fdf-0843645a4909/bars` 
- ❌ This 404s: `GET api/foo/74180c32cc6f47b69fdf0843645a4909/bars`

So the problem is not what I thought it was: it's just that the UUID default convertor is a bit too zealous (to my taste at least :stuck_out_tongue: ) and does not accept a formatting stripped from hyphens.

The actual nested/mounts works as intended. Sorry for the noise.

**Reply by edthrn** [reply]
Just for the record, Python considers the string UUID formatting without hyphens as valid:
![image](https://github.com/user-attachments/assets/4469b690-fc61-44b6-b5b1-d726b3d8a21b)

If you want me to open a PR to "fix" that, I'd be glad to do so. Let me know.

**Reply by adriangb** [reply]
I think a PR to make UUIDs with no hyphens work would be welcome 😄

**Reply by edthrn** [reply, marked as answer]
ok! Done here #2806
